test/f90_correct/src/pure_function.f90

```diff
@@ -0,0 +1,44 @@
+module mymod
```

Comment:
Could you add a comment in this test case to indicate that this is a reproducer for #1016, and that the procedures `func` and `other_func` should never be reordered?

Reply:
Alternatively, rename the test case to `issue1016.f90`.

Also, I suggest using more unique names for the module `mymod` and the program `test`. We have run into test failures previously when module names collided.

Reply:
I wouldn't rename the file to point at github issue. Let's not create a precedence.

Reply:
Would you object to mentioning the issue inside a comment?

Reply:
Suggested changes added to the test. Also rebase for new master commits